### ISSUE TITLE
ICE with unboxed and diverging closure

### ISSUE THREAD

**Issue by nodakai** [rust-lang/rust#20105]
```rust
fn main() {
    std::task::spawn(move || panic!("A")); // OK
    std::thread::Thread::spawn(move || panic!("B")).detach(); // Can't infer the return type
}
```

- http://is.gd/46Dv8w [playpen]

This makes the deprecation process of `task::spawn()` non-trivial.

`Thread::spawn::<(), FnOnce()+Send>(move || panic!("B")).detach();` did not work, either.

**Comment by jakub-** [comment]
```rust
fn f<F, R>(_: F) where F: Fn() -> R {
}

fn main() {
    f(|| -> ! { () });
}
```

works fine now. E-needstest.
